src/Model/Join.php

```diff
@@ -167,30 +167,39 @@ protected function init(): void
     {
         $this->_init();
 
-        // handle foreign table containing a dot
+        // owner model should have id_field set
+        $id_field = $this->getOwner()->id_field;
+        if (!$id_field) {
+            throw (new Exception('Joins owner model should have id_field set'))
+                ->addMoreInfo('model', $this->getOwner());
+        }
+
+        // handle foreign table containing a dot - that will be reverse join
         if (is_string($this->foreign_table) && strpos($this->foreign_table, '.') !== false) {
+            // split by LAST dot in foreign_table name
+            [$this->foreign_table, $this->foreign_field] = preg_split('~\.+(?=[^.]+$)~', $this->foreign_table);
+
             if (!isset($this->reverse)) {
                 $this->reverse = true;
-                if (isset($this->master_field)) {
-                    // both master and foreign fields are set
-
-                    // master_field exists, no we will use that
-                    // if (!is_object($this->master_field) && !$this->getOwner()->hasField($this->master_field)) {
-                    throw (new Exception('You are trying to link tables on non-id fields. This is not implemented yet'))
-                        ->addMoreInfo('condition', $this->getOwner()->table . '.' . $this->master_field . ' = ' . $this->foreign_table);
-                    // } $this->reverse = 'link';
-                }
             }
+        }
 
-            // split by LAST dot in foreign_table name
-            [$this->foreign_table, $this->foreign_field] = preg_split('/\.+(?=[^\.]+$)/', $this->foreign_table);
+        if ($this->reverse === true) {
+            if (isset($this->master_field) && $this->master_field !== $id_field) {
```

Comment:
```suggestion
            if (isset($this->master_field) && $this->master_field !== $id_field) { // TODO not implemented yet, see https://github.com/atk4/data/pull/802
```

Reply:
This is not related to this PR. It's just to ignore case when user explicitly set master_field which is the same as default one set by ATK. See my other comment above.